### PR TITLE
Implementation of XEP-0308: Last Message Correction

### PR DESCRIPTION
**Pull request by ramabit**
## **Fixes SMACK-714**
Documentation: https://xmpp.org/extensions/xep-0308.html

**How to use it**

Send: 
`message.addExtension(new MessageCorrectExtension(idInitialMessage));`

Receive:
`ProviderManager.addExtensionProvider(MessageCorrectExtension.ELEMENT_NAME, MessageCorrectExtension.NAMESPACE, new MessageCorrectProvider());`

`MessageCorrectExtension messageCorrectExtension = MessageCorrectExtension.from(message)`

### CODE REVIEW

**Comment by ramabit** [comment]
@Flowdalic TravisCI is killing the process of the 2nd check (java 1.7). I guess it is because of low memory.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/message/correct/MessageCorrectExtension.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
+/**
+ *
+ * Copyright the original author or authors
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/message/correct/MessageCorrectExtension.java]: That line should include a copyright year, i.e. 2016.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/message/correct/MessageCorrectExtension.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+ * the XEP for more implementation guidelines.
+ * 
+ * @author Fernando Ramirez, f.e.ramirez94@gmail.com
+ */
+public class MessageCorrectExtension implements ExtensionElement {
+
+    /**
+     * The XML element name of a 'message correct' extension.
+     */
+    public static final String ELEMENT_NAME = "replace";
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/message/correct/MessageCorrectExtension.java]: Should be just 'ELEMENT'

File: smack-extensions/src/main/java/org/jivesoftware/smackx/message/correct/MessageCorrectExtension.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    /*
+     * (non-Javadoc)
+     * 
+     * @see org.jivesoftware.smack.packet.PacketExtension#toXML()
+     */
+    @Override
+    public CharSequence toXML() {
+        final StringBuilder stringBuilder = new StringBuilder();
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/message/correct/MessageCorrectExtension.java]: Please use XmlStringBuilder here, and change the return type to this.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/message/correct/MessageCorrectExtension.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        return NAMESPACE;
+    }
+
+    /**
+     * A ExtensionElementProvider for the MessageCorrectExtension. 
+     * As MessageCorrection elements have only the JID of the message to replace.
+     * 
+     * @author Fernando Ramirez, f.e.ramirez94@gmail.com
+     */
+    public static class Provider extends ExtensionElementProvider<MessageCorrectExtension> {
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/message/correct/MessageCorrectExtension.java]: Please factor the provider out in a extra class under `.provider`

File: smack-extensions/src/main/java/org/jivesoftware/smackx/message/correct/MessageCorrectExtension.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+import org.jivesoftware.smack.packet.ExtensionElement;
+import org.jivesoftware.smack.provider.ExtensionElementProvider;
+import org.xmlpull.v1.XmlPullParser;
+import org.xmlpull.v1.XmlPullParserException;
+
+/**
+ * An Extension that implements XEP-0308: Last Message Correction
+ * 
+ * This extension is expected to be added to message stanzas. Please refer to
+ * the XEP for more implementation guidelines.
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/message/correct/MessageCorrectExtension.java]: You should at least mention *which* XEP you are talking about. An `<a/>` tag in `@see` couldn't hurt either.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/message/correct/MessageCorrectExtension.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+    /**
+     * The id tag of a 'message correct' extension.
+     */
+    private static final String JID_TAG = "id";
+
+    /**
+     * The jid of the message to correct.
+     */
+    private String jidInitialMessage;
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/message/correct/MessageCorrectExtension.java]: That's no JID/XMPP address, this is just the message ID. Please rename this to `id`.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/message/correct/package-info.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Classes and Interfaces that implement Last Message Correction as defined in XEP-0308.
+ */
+package org.jivesoftware.smackx.message.correct;
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/message/correct/package-info.java]: Please change the package name to `message_correct` (if it is possible, but I think it should).

File: smack-extensions/src/test/java/org/jivesoftware/smackx/message/correct/MessageCorrectExtensionTest.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+    private final CharSequence messageCorrectionXml = "<replace id='bad1' xmlns='urn:xmpp:message-correct:0'/>";
+
+    private final CharSequence expectedXml = "<message to='juliet@capulet.net/balcony' id='good1'>"
+            + "<body>But soft, what light through yonder window breaks?</body>"
+            + "<replace id='bad1' xmlns='urn:xmpp:message-correct:0'/>" + "</message>";
+
+    @Before
+    public void setStanzas() throws Exception {
+        initialMessage = (Message) PacketParserUtils.parseStanza(initialMessageXml);
```
Comment (Flowdalic) [inline smack-extensions/src/test/java/org/jivesoftware/smackx/message/correct/MessageCorrectExtensionTest.java]: That's a strange approach. Smack usually does the parsing in the `@Test` method. What's the advantage of having it done in a `@Before` method?

File: smack-extensions/src/main/java/org/jivesoftware/smackx/message/correct/MessageCorrectExtension.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    /**
+     * The namespace that qualifies the XML element of a 'message correct'
+     * extension.
+     */
+    public static final String NAMESPACE = "urn:xmpp:message-correct:0";
+
+    /**
+     * The id tag of a 'message correct' extension.
+     */
+    private static final String JID_TAG = "id";
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/message/correct/MessageCorrectExtension.java]: Same here, this is tag is not related to JIDs at all.

**Comment by Flowdalic** [comment]
Please squash merge all commits into a single commit and make sure that the issue's key 'SMACK-714' is mentioned somewhere in the commit message's body. Please don't put it into the title. Something like "Fixes SMACK-714" is sufficient.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/message/correct/MessageCorrectExtension.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+/**
+ * An Extension that implements XEP-0308: Last Message Correction
+ * 
+ * This extension is expected to be added to message stanzas. Please refer to
+ * the XEP for more implementation guidelines.
+ * 
+ * @author Fernando Ramirez, f.e.ramirez94@gmail.com
+ */
+public class MessageCorrectExtension implements ExtensionElement {
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/message/correct/MessageCorrectExtension.java]: It's Smackomatic (a Smack Idiom) that ExtensionElements have a static `from()` method. Since XEP-308 is only defined for messages, in this case we should have

`static MessageCorrectionExtension from(Message)`

Please add this method.

**Comment by ramabit** [comment]
@Flowdalic Thanks for the feedback ! 
The corrections are done and the pull request description is updated.
TravisCI is failing because a test that I didn't touch. Also, I can't reproduce that failure locally.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/message_correct/MessageCorrectExtension.java [hunk 11]
```diff
@@ -83,11 +81,13 @@ public String getElementName() {
      * @see org.jivesoftware.smack.packet.PacketExtension#toXML()
      */
     @Override
-    public CharSequence toXML() {
-        final StringBuilder stringBuilder = new StringBuilder();
-        stringBuilder.append('<').append(getElementName()).append(' ').append(JID_TAG).append('=').append('\'')
-                .append(getJidInitialMessage()).append('\'').append(" xmlns='").append(getNamespace()).append("'/>");
-        return stringBuilder.toString();
+    public XmlStringBuilder toXML() {
+        XmlStringBuilder xml = new XmlStringBuilder(this, getNamespace());
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/message_correct/MessageCorrectExtension.java]: Just `new XmlStringBuilder(this)` will do the right thing, i.e. adding the opening tag and the xmlns attribute.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/message_correct/MessageCorrectExtension.java [hunk 12]
```diff
@@ -83,11 +81,13 @@ public String getElementName() {
...
+        xml.attribute(ID_TAG, getIdInitialMessage());
+        xml.attribute("xmlns", getNamespace());
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/message_correct/MessageCorrectExtension.java]: This would create a duplicate `xmlns` attribute and thus invalid XML.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/message_correct/MessageCorrectExtension.java [hunk 13]
```diff
@@ -83,11 +81,13 @@ public String getElementName() {
...
+        xml.append('/');
+        xml.rightAngleBracket();
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/message_correct/MessageCorrectExtension.java]: Instead of these two, please use `closeEmptyElement()` instead.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/message_correct/MessageCorrectExtension.java [hunk 14]
```diff
@@ -100,21 +100,8 @@ public String getNamespace() {
...
-
-        @Override
-        public MessageCorrectExtension parse(XmlPullParser parser, int initialDepth)
-                throws XmlPullParserException, IOException, SmackException {
-            String jidMessageToReplace;
-            jidMessageToReplace = parser.getAttributeValue("", JID_TAG);
-            return new MessageCorrectExtension(jidMessageToReplace);
-        }
+    public static MessageCorrectExtension from(Message message) {
+        return new MessageCorrectExtension(message.getStanzaId());
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/message_correct/MessageCorrectExtension.java]: It appears I was unclear with what `from(Message)` should do: Instead of creating a new MessageCorrect, it should extract the extension from the message if there is one. The body of the function should be
`return message.getExtension(ELEMENT, NAMESPACE);`

File: smack-extensions/src/main/java/org/jivesoftware/smackx/message_correct/provider/MessageCorrectProvider.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    /**
+     * The id tag of a 'message correct' extension.
+     */
+    private static final String ID_TAG = "id";
+
+    @Override
+    public MessageCorrectExtension parse(XmlPullParser parser, int initialDepth)
+            throws XmlPullParserException, IOException, SmackException {
+        String jidMessageToReplace;
+        jidMessageToReplace = parser.getAttributeValue("", ID_TAG);
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/message_correct/provider/MessageCorrectProvider.java]: Instead of defining ID_TAG again, you could simply reuse ID_TAG from MessageCorrectExtension by making it public.

**Comment by Flowdalic** [comment]
> TravisCI is failing because a test that I didn't touch. Also, I can't reproduce that failure locally.

Some older Smack unit tests are not properly synchronized causing them to behave non deterministc especially on slow systems like the Travis CI infrastructure. I've restarted Travis (not sure if you are also able to hit the restart button, e.g. at https://travis-ci.org/igniterealtime/Smack/jobs/122258804 in the upper right corner).

Looks good so far :+1: I've added a few last remarks, please incorporate the changes, then squash merge all you commits into a single commit and then force push to the source branch of this PR.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/message_correct/MessageCorrectExtension.java (hunk 12 above)
Comment (ramabit) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/message_correct/MessageCorrectExtension.java]: But I want to show the id attribute before the xmlns attribute. I'll do this
`        XmlStringBuilder xml = new XmlStringBuilder();
        xml.halfOpenElement(ELEMENT);
        xml.attribute(ID_TAG, getIdInitialMessage());
        xml.xmlnsAttribute(NAMESPACE);
        xml.closeEmptyElement();`

**Comment by Flowdalic** [comment]
By the way, this PR only implements the low-level API parts for XEP-308. While it's ok to start with them first, do you plan to submit an subsequent PR with the high-level API, i.e. the MessageCorrectionManager)?

**Comment by ramabit** [comment]
> By the way, this PR only implements the low-level API parts for XEP-308. While it's ok to start with them first, do you plan to submit an subsequent PR with the high-level API, i.e. the MessageCorrectionManager)? 
> 

It's not planned up to now, but I think we can discuss it.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/message_correct/MessageCorrectExtension.java (hunk 12 above)
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/message_correct/MessageCorrectExtension.java]: > But I want to show the id attribute before the xmlns attribute.

Why? The order is completely irrelevant, but it's common practice to have the unprefixed `xmlns` attribute first (if one is required). This is why the XmlStringBuilder API is designed this way.

Comment (ramabit) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/message_correct/MessageCorrectExtension.java]: Ok :)

**Comment by ramabit** [comment]
@Flowdalic Changes done

**Comment by Flowdalic** [comment]
Perfect, now you need to squash the commits into one.

**Comment by ramabit** [comment]
@Flowdalic Github has a new option to do that when merge.

**Comment by ramabit** [comment]
@Flowdalic I can close this PR an open another one squashing the commits

**Comment by Flowdalic** [comment]
> I can close this PR an open another one squashing the commits

Simply update **this** PR by updating its source branch with a forced push.

**Comment by ramabit** [comment]
Simply update this PR by updating its source branch with a forced push.
> 

@Flowdalic done

**Comment by Flowdalic** [comment]
Thanks, but the message's body now full of meaningless half sentences, it also does not contain the issue key, which is required so that JIRA connects the issue with this commit (https://github.com/igniterealtime/Smack/pull/73#issuecomment-207594222). Please fix that.

**Comment by ramabit** [comment]
@Flowdalic like that?

**Comment by Flowdalic** [comment]
Picked as https://github.com/igniterealtime/Smack/commit/ca58c65a26034bd2b7c4b7e8c2bc70ab593d2ec7

Thanks again :)
